query-planner-js/src/buildPlan.ts

```diff
@@ -763,12 +764,74 @@ class FetchGroup {
     return this._inputs?.clone();
   }
 
-  addDependencyOn(groups: FetchGroup | FetchGroup[], pathInGroups?: OperationPath) {
-    this.dependencyGraph.addDependency(this, groups, pathInGroups);
+  addParents(parents: readonly ParentRelation[]) {
+    this.dependencyGraph.addParents(this, parents);
   }
 
-  removeDependencyOn(groups: FetchGroup | FetchGroup[]) {
-    this.dependencyGraph.removeDependency(this, groups);
+  addParent(parent: ParentRelation) {
+    this.dependencyGraph.addParent(this, parent);
+  }
+
+  removeChild(child: FetchGroup) {
+    this.dependencyGraph.removeChild(this, child);
+  }
+
+  isParentOf(maybeChild: FetchGroup): boolean {
+    return this.dependencyGraph.isChild(maybeChild, this);
+  }
+
+  isChildOf(maybeParent: FetchGroup): boolean {
+    return !!this.parentRelation(maybeParent);
+  }
+
+  /**
+   * Returns whether this group is both a child of `maybeParent` but also if we can show that the
+   * dependency between the group is "artificial" in the sense that this group inputs do not truly
+   * depend on anything `maybeParent` fetches.
+   */
+  isChildOfWithArtificialDependency(maybeParent: FetchGroup): boolean {
+    const relation =  this.parentRelation(maybeParent);
+    // To be a child with an artificial dependency, it needs to be a child first,
+    // and the "path in parent" should be know to be empty (which essentially
+    // means the groups are on the same entity at the same level).
+    if (!relation || relation.path?.length !== 0) {
```

Comment:
> Is that the behavior you want?

It is. And I was hoping that the comment was clear enough that this is what we want as I'm not sure how to explain it much better. But to rephrase that comment, we want to restrict ourselves to the case where the child and parent fetches are on the same entity/same level (it could be we can relax that restriction, but I'm not sure if it's even useful and haven't though deeply about what that involves), and `path === 0` is what tells us that. `path > 0` would tell us that the child is fetching things at a deeper level than the parent, while `path === undefined` essentially means "don't make any assumption" (that last path may be where the confusion is coming from? If so, I did added precision in the comment on `ParentRelation`).